pkg/git/poll/handler.go

```diff
@@ -24,70 +31,96 @@ type Watcher interface {
 
 // Handler handles all the watches for the git repositories. These watches are pulling the latest commit every syncPeriod.
 type Handler struct {
-	client      client.Client
-	watches     map[string]Watcher
-	createWatch func(gitRepo v1alpha1.GitRepo, client client.Client) Watcher // this func creates a watch. It's a struct field, so it can be replaced for a mock in unit tests.
-	log         logr.Logger
+	client    client.Client
+	log       logr.Logger
+	scheduler quartz.Scheduler
+	fetcher   GitFetcher
 }
 
-func NewHandler(client client.Client) *Handler {
+func NewHandler(ctx context.Context, client client.Client) *Handler {
+	scheduler := quartz.NewStdScheduler()
+	scheduler.Start(ctx)
 	return &Handler{
-		client:      client,
-		watches:     make(map[string]Watcher),
-		createWatch: NewWatch,
-		log:         ctrl.Log.WithName("git-latest-commit-poll-handler"),
+		client:    client,
+		log:       ctrl.Log.WithName("git-latest-commit-poll-handler"),
+		scheduler: scheduler,
+		fetcher:   &git.Fetch{},
 	}
 }
 
-// AddOrModifyGitRepoWatch adds a new watch for the gitrepo if no watch was already present.
-// It updates the existing watch for this gitrepo if present.
-func (h *Handler) AddOrModifyGitRepoWatch(ctx context.Context, gitRepo v1alpha1.GitRepo) {
-	key := getKey(gitRepo)
-	watch, found := h.watches[key]
-	if !found {
-		h.watches[key] = h.createWatch(gitRepo, h.client)
-		h.watches[key].StartBackgroundSync(ctx)
+// AddOrModifyGitRepoPollJob adds a new scheduled job for the gitrepo if no job was already present.
+// It updates the existing job for this gitrepo if present.
+func (h *Handler) AddOrModifyGitRepoPollJob(ctx context.Context, gitRepo v1alpha1.GitRepo) {
+	gitRepoPollKey := GitRepoPollKey(gitRepo)
+	scheduledJob, err := h.scheduler.GetScheduledJob(gitRepoPollKey)
+	if err != nil {
+		// job was not found
+		if gitRepo.Spec.DisablePolling {
+			// nothing to do if disablePolling is set
+			return
+		}
+		h.scheduleJob(ctx, gitRepoPollKey, gitRepo, true)
 	} else {
-		oldSyncInterval := watch.GetSyncInterval()
-		watch.UpdateGitRepo(gitRepo)
-
-		gitRepoSyncInterval := 0.0
-		if pi := gitRepo.Spec.PollingInterval; pi != nil {
-			gitRepoSyncInterval = pi.Seconds()
+		if gitRepo.Spec.DisablePolling {
+			// if polling is disabled, just delete the job from the scheduler
+			err = h.scheduler.DeleteJob(gitRepoPollKey)
+			if err != nil {
+				h.log.Error(err, fmt.Sprintf("Error deleting job %s", gitRepoPollKey))
```

Comment:
This might be a problem, the signature is
`func (l logr.Logger) Error(err error, msg string, keysAndValues ...any)`